gufe/protocols/protocol.py

```diff
@@ -0,0 +1,89 @@
+# This code is part of OpenFE and is licensed under the MIT license.
+# For details, see https://github.com/OpenFreeEnergy/gufe
+
+import abc
+from typing import Optional
+
+from pydantic import BaseModel, validator
+from openff.toolkit.utils.serialization import Serializable
+
+from ..chemicalsystem import ChemicalSystem
+from ..mapping import AtomMapping
+from .protocol_settings import ProtocolSettings
+
+
+class Protocol(abc.ABC, Serializable):
+    """A protocol that implements an alchemical transformation.
+
+    Takes a `ProtocolSettings` object specific to the protocol on init.
+    This configures the protocol for repeated execution on `ChemicalSystem`s.
+
+    The `run` method takes the `start` and `end` `ChemicalSystem`,
+    the `AtomMapping`, and any additional keywords specific to the `Protocol`.
+
+    Attributes
+    ----------
+    settings : 
+
+    """
+    ...
+
+    def __init__(
+            self,
+            *,
+            settings: ProtocolSettings = None
+        ):
+        """
```

Comment:
If the two `ChemicalSystems` aren't here, then this class becomes a factory for simulations.  I think being able to serialize the factory separately is a good idea for extending a network with the same protocol (else where is the factory stored).

In which case, I think `run()` isn't run, but instead `prepare(Transformation(initial, final, mapping)) -> PreparedProtocol` and `PreparedProtocol` has the `run()` and `is_complete()` etc.

Reply:
Indeed, here are my notes when I separated things out this way:

2. Benefits of separating `Transformation` from `Protocol`
    - can have a single `Protocol` object instance serving many `Transformation`s
        - can save on memory for very large networks, serialization cost
        - may be worthwhile making them function as singletons when the hash is identical
    - only two types of `Transformation`: `Transformation` and `NonTransformation`
        - we allow for a zoo of `Protocol`s
    - can execute a `Protocol` in isolation, without any graph/network machinery at all
        - define `ChemicalSystem`s, `AtomMapping` if appropriate, settings, and fire it off with 
          ```
          Protocol(settings).run(start, end, mapping)
          ```
          
         - very similar to `sklearn`-style estimator classes
         - should we adopt a similar approach to `sklearn` for results?

Reply:
Will have to think some more on `Protocol` structure to accommodate @jchodera's [points here](https://github.com/OpenFreeEnergy/gufe/pull/13#discussion_r853994681). I think you're right that a `Protocol` with an `initial`, `final`, and `mapping` at this point is a job that's ready to be executed, and it is what would need to be serializable to be shipped to a worker process, local or remote.

`PreparedProtocol.run` could have an `async=True` kwarg that proceeds with execution in a separate thread or process, with its methods able to yield result status from file outputs or stream buffers to the main thread/process. We'll need to determine what operations we support (stop, start, extend?).